doc/tutorial/read_mapper/read_mapper_step3.cpp

```diff
@@ -0,0 +1,109 @@
+#include <seqan3/argument_parser/all.hpp>
+#include <seqan3/io/sequence_file/input.hpp>
+#include <seqan3/io/stream/debug_stream.hpp>
+#include <seqan3/range/view/get.hpp>
+#include <seqan3/search/algorithm/all.hpp>
+#include <range/v3/view/zip.hpp>
+
+using namespace seqan3;
+
+struct data_storage_t
+{
+    concatenated_sequences<std::string> ref_id;
+    concatenated_sequences<std::vector<dna5>> ref_seq;
+    concatenated_sequences<std::string> query_id;
+    concatenated_sequences<std::vector<dna5>> query_seq;
+    concatenated_sequences<std::vector<phred42>> query_qual;
+};
+
+//! [solution]
+void read_input(std::filesystem::path const & reference_path,
+                std::filesystem::path const & query_path,
+                data_storage_t & storage)
+{
+    sequence_file_input reference_in{reference_path};
+    storage.ref_id  = std::move(get<field::ID>(reference_in));
+    storage.ref_seq = std::move(get<field::SEQ>(reference_in));
+
+    sequence_file_input query_in{query_path};
+    storage.query_id   = std::move(get<field::ID>(query_in));
+    storage.query_seq  = std::move(get<field::SEQ>(query_in));
+    storage.query_qual = std::move(get<field::QUAL>(query_in));
+}
+
+//! [map]
+void map(uint8_t const errors, data_storage_t & storage)
+//! [map]
+{
+    bi_fm_index index{storage.ref_seq};
+
+    configuration const search_config = search_cfg::max_error{search_cfg::total{errors}} |
+                                        search_cfg::mode{search_cfg::all_best};
+
+    for (auto && [id, query, quality] : ranges::view::zip(storage.query_id, storage.query_seq, storage.query_qual) |
+                                        view::take(20))
```

Comment:
instead of zipping here, you could just stream over the query file, wouldn't that be easier?

Reply:
Since we only do it once... yes 😅